java-squid/src/main/java/org/sonar/java/model/DefaultJavaFileScannerContext.java

```diff
@@ -0,0 +1,230 @@
+
+  public DefaultJavaFileScannerContext(
+    CompilationUnitTree tree, SourceFile sourceFile, File file, SemanticModel semanticModel, boolean analyseAccessors, @Nullable SonarComponents sonarComponents) {
+    this.tree = tree;
+    this.sourceFile = sourceFile;
+    this.file = file;
+    this.semanticModel = semanticModel;
+    this.sonarComponents = sonarComponents;
+    this.complexityVisitor = new ComplexityVisitor(analyseAccessors);
+  }
+
+  @Override
+  public CompilationUnitTree getTree() {
+    return tree;
+  }
+
+  @Override
+  public void addIssue(Tree tree, JavaCheck javaCheck, String message) {
+    addIssue(((JavaTree) tree).getLine(), javaCheck, message, null);
+  }
+
+  @Override
+  public void addIssue(Tree tree, JavaCheck check, String message, @Nullable Double cost) {
+    addIssue(((JavaTree) tree).getLine(), check, message, cost);
+  }
+
+  @Override
+  public void addIssueOnFile(JavaCheck javaCheck, String message) {
+    addIssue(-1, javaCheck, message);
+  }
+
+  @Override
+  public void addIssue(int line, JavaCheck javaCheck, String message) {
+    addIssue(line, javaCheck, message, null);
+  }
+
+  @Override
+  public void addIssue(int line, JavaCheck javaCheck, String message, @Nullable Double cost) {
+    Preconditions.checkNotNull(javaCheck);
+    Preconditions.checkNotNull(message);
+    RuleKey key = getRuleKey(javaCheck);
+    if (key != null) {
+      Issuable issuable = sonarComponents.issuableFor(file);
+      if (issuable != null) {
+        Issuable.IssueBuilder issueBuilder = issuable.newIssueBuilder()
+          .ruleKey(key)
+          .message(message);
+        if (line > 0) {
+          issueBuilder.line(line);
+        }
+        if (cost == null) {
+          Annotation linear = AnnotationUtils.getAnnotation(javaCheck, SqaleLinearRemediation.class);
+          Annotation linearWithOffset = AnnotationUtils.getAnnotation(javaCheck, SqaleLinearWithOffsetRemediation.class);
+          if (linear != null || linearWithOffset != null) {
+            throw new IllegalStateException("A check annotated with a linear sqale function should provide an effort to fix");
+          }
+        } else {
+          issueBuilder.effortToFix(cost);
+        }
+        Issue issue = issueBuilder.build();
+        issuable.addIssue(issue);
+      }
+    }
+  }
+
+  @Override
+  @Nullable
+  public Object getSemanticModel() {
+    return semanticModel;
+  }
+
+  @Override
+  public String getFileKey() {
+    return sourceFile.getKey();
+  }
+
+  @CheckForNull
+  private RuleKey getRuleKey(JavaCheck check) {
+    if (sonarComponents != null) {
+      return sonarComponents.getRuleKey(check);
+    }
+    return null;
+  }
+
+  @Override
+  public void addIssue(File file, JavaCheck check, int line, String message) {
+    if (sonarComponents != null) {
+      sonarComponents.addIssue(file, check, line, message);
+    }
+  }
+
+  /**
+   * FIXME(mpaladin) DO NOT GO ON RELEASE WITH THIS CONSTANT SET TO TRUE
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)